File: src/main/java/com/example/coopachat/services/admin/AdminService.java
package com.example.coopachat.services.admin;

import com.example.coopachat.enums.SupplierType;

import com.example.coopachat.dtos.delivery.DeliveryOptionDTO;
import com.example.coopachat.dtos.fee.CreateFeeDTO;
import com.example.coopachat.dtos.fee.FeeDTO;
import com.example.coopachat.dtos.categories.CategoryListItemDTO;
import com.example.coopachat.dtos.categories.CategoryKpiDTO;
import com.example.coopachat.dtos.categories.CreateCategoryDTO;
import com.example.coopachat.dtos.categories.UpdateCategoryDTO;
import com.example.coopachat.dtos.products.CreateProductDTO;
import com.example.coopachat.dtos.products.ProductDetailsDTO;
import com.example.coopachat.dtos.products.ProductListResponseDTO;
import com.example.coopachat.dtos.products.UpdateProductDTO;
import com.example.coopachat.dtos.products.ProductStatsDTO;
import com.example.coopachat.dtos.products.TopProductUsageDTO;
import com.example.coopachat.dtos.products.UpdateProductStatusDTO;
import com.example.coopachat.dtos.user.SaveUserDTO;
import com.example.coopachat.dtos.user.UpdateUserStatusDTO;
import com.example.coopachat.dtos.user.UserDetailsDTO;
import com.example.coopachat.dtos.user.UserListResponseDTO;
import com.example.coopachat.dtos.user.UserStatsByRoleItemDTO;
import com.example.coopachat.dtos.user.UserStatsByStatusItemDTO;
import com.example.coopachat.dtos.user.UserStatsDTO;
import com.example.coopachat.dtos.suppliers.*;
import com.example.coopachat.dtos.dashboard.admin.AdminAlertsDTO;
import com.example.coopachat.dtos.dashboard.admin.AdminDashboardStatsDTO;
import com.example.coopachat.dtos.dashboard.admin.CouponUsageParJourDTO;
import com.example.coopachat.dtos.dashboard.admin.LivraisonParJourDTO;
import com.example.coopachat.dtos.dashboard.admin.StockEtatGlobalDTO;
import com.example.coopachat.dtos.dashboard.logisticsManager.StatutTourneesDTO;
import com.example.coopachat.dtos.reference.CreateReferenceItemDTO;
import com.example.coopachat.dtos.reference.ReferenceItemDTO;
import com.example.coopachat.enums.UserRole;
import java.util.List;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.web.multipart.MultipartFile;

/**
 * Interface pour le service de gestion des actions de l'administrateur
 */
public interface AdminService {

    /**
     * Crée une nouvelle catégorie
     *
     * @param createCategoryDTO Les informations de la catégorie à créer
     * @throws RuntimeException si le nom de la catégorie existe déjà ou si une erreur survient
     */
    void createCategory(CreateCategoryDTO createCategoryDTO);

    /**
     * Récupère la liste des catégories (id + nom + icon).
     */
    List<CategoryListItemDTO> getAllCategories(String search);

    /**
     * KPI de la page catégories (total catégories, total produits, produits actifs).
     */
    CategoryKpiDTO getCategoryKpis();

    /**
     * Récupère une catégorie par son ID (id + nom + icon).
     */
    CategoryListItemDTO getCategoryById(Long id);

    /**
     * Met à jour une catégorie : seuls les champs non null du DTO sont modifiés.
     */
    void updateCategory(Long id, UpdateCategoryDTO dto);

    /**
     * Supprime une catégorie ainsi que tous les produits rattachés.
     */
    void deleteCategory(Long id);

    /**
     * Crée un nouveau produit
     *
     * @param createProductDTO Les informations du produit à créer
     * @throws RuntimeException si le nom du produit existe déjà, si la catégorie n'existe pas ou si une erreur survient
     */
    void createProduct(CreateProductDTO createProductDTO);

    /**
     * Récupère la liste paginée de tous les produits avec recherche et filtres
     *
     * @param page Numéro de la page (0-indexed)
     * @param size Taille de la page
     * @param search Terme de recherche (nom ou code produit)
     * @param categoryId ID de la catégorie pour filtrer
     * @param status Statut actif/inactif pour filtrer (true = actif, false = inactif)
     * @return Réponse paginée contenant la liste des produits
     */
    ProductListResponseDTO getAllProducts(int page, int size, String search, Long categoryId, Boolean status);

    /**
     * Récupère les détails d'un produit par son ID
     *
     * @param id ID du produit
     * @return DTO contenant les détails du produit
     * @throws RuntimeException si le produit n'existe pas ou si une erreur survient
     */
    ProductDetailsDTO getProductById(Long id);

    /**
     * Met à jour un produit existant
     *
     * @param id ID du produit à modifier
     * @param updateProductDTO Les informations à mettre à jour (tous les champs sont optionnels)
     * @throws RuntimeException si le produit n'existe pas, si le nom existe déjà, si la catégorie n'existe pas ou si une erreur survient
     */
    void updateProduct(Long id, UpdateProductDTO updateProductDTO);

    /**
     * Active ou désactive un produit
     *
     * @param id ID du produit à activer/désactiver
     * @param updateProductStatusDTO Le statut à appliquer (true = actif, false = inactif)
     * @throws RuntimeException si le produit n'existe pas ou si une erreur survient
     */
    void updateProductStatus(Long id, UpdateProductStatusDTO updateProductStatusDTO);

    /**
     * Exporte la liste des produits en fichier Excel selon les filtres.
     * Le service retourne les données brutes (ByteArrayResource).
     * Le controller ajoute les headers HTTP (Content-Disposition, Content-Type) pour le téléchargement et retourne ResponseEntity<Resource> (la réponse complète).
     *
     * @param search Terme de recherche (nom ou code produit)
     * @param categoryId ID de la catégorie pour filtrer
     * @param status Statut actif/inactif pour filtrer (true = actif, false = inactif)
     * @return ByteArrayResource contenant les données brutes du fichier Excel
     * @throws RuntimeException si une erreur survient lors de la génération du fichier
     */
    ByteArrayResource exportProducts(String search, Long categoryId, Boolean status);

    /**
     * Récupère les statistiques des produits
     *
     * @return  ProductStatsDTO  contenant total, activé, désactivé
     * @throws RuntimeException si une erreur survient
     */
    ProductStatsDTO getProductStats();

    /**
     * Récupère le top 5 des produits les plus commandés avec leur taux d'utilisation en %.
     * Chaque produit est retourné avec son nom et le pourcentage que représentent ses
     * quantités commandées par rapport au total des quantités sur la période (ex. 30 derniers jours).
     *
     * @return liste de TopProductUsageDTO (productName, usagePercent entre 0 et 100)
     */
    List<TopProductUsageDTO> getTop5ProductUsage();

    /**
     * Récupère la liste des fournisseurs (id + nom)
     *
     * @return Liste des fournisseurs
     */
    List<SupplierListItemDTO> getAllSuppliers();

    /**
     * Crée un nouveau fournisseur avec toutes les informations détaillées.
     *
     * @param dto Les informations du fournisseur à créer
     */
    void createSupplier(CreateSupplierDTO dto);

    /**
     * Liste paginée des fournisseurs avec recherche et filtres.
     */
    SupplierListResponseDTO getSuppliers(int page, int size, String search, Long categoryId, SupplierType type, Boolean status);

    /**
     * Récupère les détails d'un fournisseur par son ID.
     */
    SupplierDetailsDTO getSupplierById(Long id);

    /**
     * Met à jour les informations d'un fournisseur.
     */
    void updateSupplier(Long id, UpdateSupplierDTO dto);

    /**
     * Active ou désactive un fournisseur.
     */
    void updateSupplierStatus(Long id, UpdateSupplierStatusDTO dto);

    /**
     * Récupère les statistiques des fournisseurs (Total, Actifs, Inactifs).
     */
    SupplierStatsDTO getSupplierStats();

    /**
     * Crée une nouvelle option de livraison
     *
     * @param dto Les informations de l'option de livraison à créer
     * @throws RuntimeException si une option avec le même nom existe déjà ou si une erreur survient
     */
    void createDeliveryOption(DeliveryOptionDTO dto);

    /**
     * Récupère toutes les options de livraison
     *
     * @return Liste des options de livraison
     */
    List<DeliveryOptionDTO> getAllDeliveryOptions();

    // ============================================================================
    // 💰 FRAIS (paramétrables par l'admin)
    // ============================================================================

    List<FeeDTO> getAllFees();
    void createFee(CreateFeeDTO dto);

    /**
     * Crée un nouvel utilisateur "agent": Admin, Responsable logistique, Commercial ou Livreur.
     * Les salariés (EMPLOYEE) ne se créent pas ici — ils sont créés par les commerciaux.
     * Envoi d'un code d'activation par email  si nécessaire .
     *
     * @param dto Données du nouvel utilisateur (prénom, nom, email, téléphone, rôle)
     */
    void createUser(SaveUserDTO dto);

    /**
     * Liste paginée de tous les utilisateurs (tous rôles) avec filtres optionnels.
     *
     * @param page   numéro de page (0-based)
     * @param size   nombre d'éléments par page
     * @param search recherche sur prénom, nom ou email (optionnel)
     * @param role   filtre par rôle (optionnel)
     * @param status filtre par statut actif/inactif (optionnel, true = actif, false = inactif)
     */
    UserListResponseDTO getUsers(int page, int size, String search, UserRole role, Boolean status);

    /**
     * Export Excel de la liste des utilisateurs (mêmes filtres que {@link #getUsers} sans pagination).
     */
    ByteArrayResource exportUsers(String search, UserRole role, Boolean status);

    /**
     * Statistiques utilisateurs pour la page Gestion des utilisateurs (total, actifs, inactifs).
     */
    UserStatsDTO getUsersStats();

    /**
     * Statistiques utilisateurs par rôle (effectif et % par rôle) pour le graphique "Utilisateurs par rôle".
     */
    List<UserStatsByRoleItemDTO> getUsersStatsByRole();

    /**
     * Répartition des statuts (Actifs / Inactifs) pour le graphique "Répartition des statuts" (donut).
     */
    List<UserStatsByStatusItemDTO> getUsersStatsByStatus();

    /**
     * Voir détails d'un utilisateur par son id .
     */
    UserDetailsDTO getUserById(Long id);

    /**
     * Activer ou désactiver un utilisateur (toggle statut).
     */
    void updateUserStatus(Long id, UpdateUserStatusDTO dto);

    /**
     * Modifier les infos de base d'un utilisateur (prénom, nom, email, téléphone, rôle, companyCommercial).
     */
    void updateUser(Long id, SaveUserDTO dto);

    /**
     * Met à jour la photo de profil d'un utilisateur (upload du fichier puis mise à jour de profilePhotoUrl).
     * Réservé à l'admin ou à l'utilisateur lui-même selon l'endpoint appelé.
     */
    void updateUserProfilePhoto(Long userId, MultipartFile file);

    /**
     * Supprime la photo de profil d'un utilisateur (fichier MinIO + champ en base).
     * Réservé à l'administrateur.
     */
    void removeUserProfilePhoto(Long userId);

    /**
     * Met à jour la photo de profil de l'utilisateur connecté (salarié, livreur, admin).
     * Utilisé par PUT /api/users/me/profile-photo.
     */
    void updateProfilePhotoForCurrentUser(MultipartFile file);

    /**
     * Supprime la photo de profil de l'utilisateur authentifié (même effet que {@link #removeUserProfilePhoto(Long)} sans être admin).
     */
    void removeProfilePhotoForCurrentUser();

    // ============================================================================
    // 📊 DASHBOARD ADMIN
    // ============================================================================

    /**
     * Statistiques du tableau de bord admin (KPIs + paiements par statut).
     * Données globales, sans filtre de période.
     *
     * @param periode ignoré (conservé pour compatibilité signature)
     * @return DTO avec : commandes en attente, paiements échoués, réclamations ouvertes, liste paiements par statut
     */
    AdminDashboardStatsDTO getDashboardStats(String periode);

    /**
     * Alias historique : même résultat que {@link #getLivraisonsParJour()} (date prévue, livrées à la date, retard).
     * GET /admin/dashboard/commandes-vs-livraisons — conservé pour compatibilité clients.
     */
    List<LivraisonParJourDTO> getCommandesVsLivraisons();

    /**
     * Effectifs pour le donut "Stocks - État global" : Normal, Sous seuil, Critique (rupture).
     */
    StockEtatGlobalDTO getStockEtatGlobal();

    /**
     * Effectifs des tournées par statut (ASSIGNEE, EN_COURS, TERMINEE, ANNULEE) pour le donut « Statut des livraisons ».
     */
    StatutTourneesDTO getStatutTournees();

    /**
     * Graphique « Livraisons 7 jours » : pour chaque des 7 derniers jours, date + nbPrevues, nbLivreesALaDate, nbRetard.
     */
    List<LivraisonParJourDTO> getLivraisonsParJour();

    /**
     * Graphique "Tendance des coupons utilisés" : pour chacun des 7 derniers jours, date (dd/MM) et nombre d'utilisations (commandes avec coupon).
     */
    List<CouponUsageParJourDTO> getCouponsUtilisesParJour();

    /**
     * Liste des alertes pour le tableau de bord admin (livraisons en retard, stocks sous seuil).
     * Chaque alerte contient type, message, detail, module (LIVRAISONS / STOCKS pour la navigation), date.
     */
    AdminAlertsDTO getAlerts();

    // ============================================================================
    // 📋 RÉFÉRENTIELS (types réclamation, raisons livraison livreur/salarié)
    // ============================================================================

    List<ReferenceItemDTO> getAllClaimProblemTypes();
    ReferenceItemDTO getClaimProblemTypeById(Long id);
    void createClaimProblemType(CreateReferenceItemDTO dto);
    void updateClaimProblemType(Long id, CreateReferenceItemDTO dto);
    void deleteClaimProblemType(Long id);

    List<ReferenceItemDTO> getAllDeliveryIssueReasons();
    ReferenceItemDTO getDeliveryIssueReasonById(Long id);
    void createDeliveryIssueReason(CreateReferenceItemDTO dto);
    void updateDeliveryIssueReason(Long id, CreateReferenceItemDTO dto);
    void deleteDeliveryIssueReason(Long id);

    List<ReferenceItemDTO> getAllEmployeeDeliveryIssueReasons();
    ReferenceItemDTO getEmployeeDeliveryIssueReasonById(Long id);
    void createEmployeeDeliveryIssueReason(CreateReferenceItemDTO dto);
    void updateEmployeeDeliveryIssueReason(Long id, CreateReferenceItemDTO dto);
    void deleteEmployeeDeliveryIssueReason(Long id);

    List<ReferenceItemDTO> getAllCompanySectors();
    ReferenceItemDTO getCompanySectorById(Long id);
    void createCompanySector(CreateReferenceItemDTO dto);
    void updateCompanySector(Long id, CreateReferenceItemDTO dto);
    void deleteCompanySector(Long id);
}

File: src/app/shared/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';


/**
 * Un élément "Paiements par statut" renvoyé par l'API (ex. Payé: 10, Échoué: 1).
 */
export interface PaymentStatusItemDTO {
  statusLabel: string;
  count: number;
}

/**
 * Réponse de GET /api/admin/dashboard/stats (données globales, sans filtre de période).
 * Les 3 KPIs + la liste pour le graphique "Paiements par statut".
 */
export interface AdminDashboardStatsDTO {
  commandesEnAttente: number;
  paiementsEchoues: number;
  reclamationsOuvertes: number;
  paiementsParStatut: PaymentStatusItemDTO[];
}

/** Un jour du graphique "Tendance des coupons utilisés" (7 derniers jours). */
export interface CouponUsageParJourDTO {
  date: string;
  nbUtilisations: number;
}

/** Une alerte du tableau de bord admin (GET /api/admin/alerts). */
export interface AlertItemDTO {
  type: string;
  message: string;
  detail: string;
  module: string;
  date: string;
}

export interface AdminAlertsDTO {
  alerts: AlertItemDTO[];
}

/**
 * Un jour du graphique « Livraisons » (7 derniers jours) : prévu à la date, livré à la date prévue, retard.
 * GET /admin/dashboard/livraisons-par-jour — et GET .../commandes-vs-livraisons (même JSON, alias).
 */
export interface LivraisonParJourDTO {
  date: string;
  nbPrevues: number;
  nbLivreesALaDate: number;
  nbRetard: number;
}

/** Un rôle du graphique "Utilisateurs par rôle" (API GET /admin/users/stats/by-role). */
export interface UserStatsByRoleItemDTO {
  role: string;
  roleLabel: string;
  count: number;
  percentage: number;
}

// --- Page Gestion des utilisateurs (GET/POST /admin/users, stats, etc.) ---

export interface UserListItemDTO {
  id: number;
  reference: string;
  firstName: string;
  lastName: string;
  email: string;
  roleLabel: string;
  profilePhotoUrl?: string | null;
  createdAt: string;
  isActive: boolean;
}

export interface UserListResponseDTO {
  content: UserListItemDTO[];
  totalElements: number;
  totalPages: number;
  currentPage: number;
  pageSize: number;
  hasNext: boolean;
  hasPrevious: boolean;
}

export interface UserStatsDTO {
  totalUsers: number;
  activeUsers: number;
  inactiveUsers: number;
}

export interface UserStatsByStatusItemDTO {
  label: string;
  count: number;
  percentage: number;
}

export interface UserDetailsDTO {
  id: number;
  refUser: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber?: string | null;
  role: string;
  roleLabel: string;
  companyCommercial?: string | null;
  companyName?: string;
  isActive: boolean;
  profilePhotoUrl?: string | null;
  createdAt: string;
}

/** Création / modification utilisateur (POST /users, PUT /users/{id}). */
export interface SaveUserDTO {
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  role: string;
  companyCommercial?: string | null;
}

/** Toggle statut (PATCH /users/{id}/status). */
export interface UpdateUserStatusDTO {
  isActive: boolean;
}

/** Réponse GET /admin/dashboard/stocks-etat-global pour le donut "Stocks - État global". */
export interface StockEtatGlobalDTO {
  normal: number;
  sousSeuil: number;
  critique: number;
}

/** Réponse GET /admin/dashboard/statut-tournees — effectifs par statut de tournée (clés ASSIGNEE, EN_COURS, …). */
export interface StatutTourneesDTO {
  parStatut: Record<string, number>;
}

// --- Référentiels / Configuration ---

export interface ReferenceItemDTO {
  id?: number;
  name: string;
  description: string;
}

export interface DeliveryOptionDTO {
  id?: number;
  name: string;
  description: string;
  isActive: boolean;
}

export interface FeeDTO {
  id?: number;
  name: string;
  description: string;
  amount: number;
  isActive: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class AdminService {
  private apiUrl = environment.apiUrl; 

  constructor(private http: HttpClient) {}

  /**
   * Récupère les stats du tableau de bord admin (3 KPIs + paiements par statut).
   * @param periode "TODAY" (aujourd'hui) ou "THIS_MONTH" (mois en cours)
   */
  getDashboardStats(): Observable<AdminDashboardStatsDTO> {
    return this.http.get<AdminDashboardStatsDTO>(`${this.apiUrl}/admin/dashboard/stats`);
  }

  /**
   * Alias historique : même réponse que getLivraisonsParJour() (nbPrevues, nbLivreesALaDate, nbRetard).
   */
  getCommandesVsLivraisons(): Observable<LivraisonParJourDTO[]> {
    return this.http.get<LivraisonParJourDTO[]>(`${this.apiUrl}/admin/dashboard/commandes-vs-livraisons`);
  }

  /**
   * 7 derniers jours : date (dd/MM), nbPrevues, nbLivreesALaDate, nbRetard.
   * Graphique « Livraisons » (empilé : livrées à la date, prévu non livré, retard).
   */
  getLivraisonsParJour(): Observable<LivraisonParJourDTO[]> {
    return this.http.get<LivraisonParJourDTO[]>(`${this.apiUrl}/admin/dashboard/livraisons-par-jour`);
  }

  /**
   * Récupère les effectifs par rôle pour le graphique "Utilisateurs par rôle".
   * GET /admin/users/stats/by-role → [{ roleLabel: "Salariés", count: 40, ... }, ...]
   */
  getUsersStatsByRole(): Observable<UserStatsByRoleItemDTO[]> {
    return this.http.get<UserStatsByRoleItemDTO[]>(`${this.apiUrl}/admin/users/stats/by-role`);
  }

  /** Liste paginée des utilisateurs (filtres optionnels). */
  getUsers(params: { page: number; size: number; search?: string; role?: string; status?: boolean }): Observable<UserListResponseDTO> {
    let httpParams = new HttpParams()
      .set('page', params.page.toString())
      .set('size', params.size.toString());
    if (params.search != null && params.search.trim() !== '') {
      httpParams = httpParams.set('search', params.search.trim());
    }
    if (params.role != null && params.role !== '') {
      httpParams = httpParams.set('role', params.role);
    }
    if (params.status != null) {
      httpParams = httpParams.set('status', params.status.toString());
    }
    return this.http.get<UserListResponseDTO>(`${this.apiUrl}/admin/users`, { params: httpParams });
  }

  /** Export Excel des utilisateurs (mêmes filtres que la liste : search, role, status). */
  exportUsers(search?: string, role?: string, status?: boolean): Observable<Blob> {
    let httpParams = new HttpParams();
    if (search != null && search.trim() !== '') {
      httpParams = httpParams.set('search', search.trim());
    }
    if (role != null && role !== '') {
      httpParams = httpParams.set('role', role);
    }
    if (status !== undefined && status !== null) {
      httpParams = httpParams.set('status', String(status));
    }
    return this.http.get(`${this.apiUrl}/admin/users/export`, {
      params: httpParams,
      responseType: 'blob'
    });
  }

  getUsersStats(): Observable<UserStatsDTO> {
    return this.http.get<UserStatsDTO>(`${this.apiUrl}/admin/users/stats`);
  }

  getUsersStatsByStatus(): Observable<UserStatsByStatusItemDTO[]> {
    return this.http.get<UserStatsByStatusItemDTO[]>(`${this.apiUrl}/admin/users/stats/by-status`);
  }

  getUserById(id: number): Observable<UserDetailsDTO> {
    return this.http.get<UserDetailsDTO>(`${this.apiUrl}/admin/users/${id}`);
  }

  updateUserStatus(id: number, dto: UpdateUserStatusDTO): Observable<string> {
    return this.http.patch(`${this.apiUrl}/admin/users/${id}/status`, dto, { responseType: 'text' });
  }

  /**
   * Met à jour un utilisateur via multipart/form-data (request params + photo optionnelle).
   * Tous les champs sont optionnels ; si profilePhoto est fourni, il remplace la photo actuelle.
   */
  updateUser(id: number, params: {
    firstName?: string;
    lastName?: string;
    email?: string;
    phoneNumber?: string;
    role?: string;
    companyCommercial?: string;
    profilePhoto?: File;
  }): Observable<string> {
    const form = new FormData();
    if (params.firstName != null) form.append('firstName', params.firstName);
    if (params.lastName != null) form.append('lastName', params.lastName);
    if (params.email != null) form.append('email', params.email);
    if (params.phoneNumber != null) form.append('phoneNumber', params.phoneNumber);
    if (params.role != null) form.append('role', params.role);
    if (params.companyCommercial != null && params.companyCommercial !== '') {
      form.append('companyCommercial', params.companyCommercial);
    }
    if (params.profilePhoto) {
      form.append('profilePhoto', params.profilePhoto, params.profilePhoto.name);
    }
    return this.http.put(`${this.apiUrl}/admin/users/${id}`, form, { responseType: 'text' });
  }

  /**
   * Crée un utilisateur via multipart/form-data (request params + photo optionnelle).
   * Paramètres : firstName, lastName, email, phoneNumber, role, companyCommercial (optionnel), profilePhoto (optionnel).
   */
  createUser(params: {
    firstName: string;
    lastName: string;
    email: string;
    phoneNumber: string;
    role: string;
    companyCommercial?: string;
    profilePhoto?: File;
  }): Observable<string> {
    const form = new FormData();
    form.append('firstName', params.firstName);
    form.append('lastName', params.lastName);
    form.append('email', params.email);
    form.append('phoneNumber', params.phoneNumber);
    form.append('role', params.role);
    if (params.companyCommercial != null && params.companyCommercial !== '') {
      form.append('companyCommercial', params.companyCommercial);
    }
    if (params.profilePhoto) {
      form.append('profilePhoto', params.profilePhoto, params.profilePhoto.name);
    }
    return this.http.post(`${this.apiUrl}/admin/users`, form, { responseType: 'text' });
  }

  /** Met à jour la photo de profil d'un utilisateur (PUT /admin/users/{id}/profile-photo). */
  updateUserProfilePhoto(userId: number, file: File): Observable<string> {
    const form = new FormData();
    form.append('file', file, file.name);
    return this.http.put(`${this.apiUrl}/admin/users/${userId}/profile-photo`, form, { responseType: 'text' });
  }

  /**
   * Supprime la photo de profil (POST /admin/users/{id}/profile-photo/remove).
   * POST évite les proxys qui ne relaient pas DELETE (erreur « Request method 'DELETE' is not supported »).
   */
  deleteUserProfilePhoto(userId: number): Observable<string> {
    return this.http.post(
      `${this.apiUrl}/admin/users/${userId}/profile-photo/remove`,
      {},
      { responseType: 'text' }
    );
  }

  /** Retourne l'URL complète pour afficher une photo de profil (ex. /api/files/profiles/xxx.jpg). */
  getProfilePhotoUrl(profilePhotoUrl: string | null | undefined): string | null {
    if (!profilePhotoUrl?.trim()) return null;
    const base = profilePhotoUrl.startsWith('http') ? '' : (environment.imageServerUrl ?? '').replace(/\/$/, '');
    return base ? `${base}/files/${profilePhotoUrl}` : `/files/${profilePhotoUrl}`;
  }

  /**
   * Récupère les effectifs Normal, Sous seuil, Critique pour le donut "Stocks - État global".
   * GET /admin/dashboard/stocks-etat-global
   */
  getStockEtatGlobal(): Observable<StockEtatGlobalDTO> {
    return this.http.get<StockEtatGlobalDTO>(`${this.apiUrl}/admin/dashboard/stocks-etat-global`);
  }

  /** Effectifs des tournées par statut — donut « Statut des livraisons ». GET /admin/dashboard/statut-tournees */
  getStatutTournees(): Observable<StatutTourneesDTO> {
    return this.http.get<StatutTourneesDTO>(`${this.apiUrl}/admin/dashboard/statut-tournees`);
  }

  /**
   * 7 derniers jours : date (dd/MM), nbUtilisations (commandes avec coupon). GET /admin/dashboard/coupons-utilises-par-jour
   */
  getCouponsUtilisesParJour(): Observable<CouponUsageParJourDTO[]> {
    return this.http.get<CouponUsageParJourDTO[]>(`${this.apiUrl}/admin/dashboard/coupons-utilises-par-jour`);
  }

  /** GET /api/admin/alerts — alertes (livraisons en retard, stocks critiques). module = LIVRAISONS | STOCKS pour la navigation. */
  getAlerts(): Observable<AdminAlertsDTO> {
    return this.http.get<AdminAlertsDTO>(`${this.apiUrl}/admin/alerts`);
  }

  // --- Catégories (id + nom + icon) ---
  /** @param noCache si true, ajoute un paramètre pour éviter le cache navigateur (après création/modification). */
  getCategories(noCache?: boolean): Observable<CategoryListItemDTO[]> {
    const options = noCache ? { params: { _: Date.now().toString() } } : {};
    return this.http.get<CategoryListItemDTO[]>(`${this.apiUrl}/admin/categories`, options);
  }

  getCategoryById(id: number): Observable<CategoryListItemDTO> {
    return this.http.get<CategoryListItemDTO>(`${this.apiUrl}/admin/categories/${id}`);
  }

  createCategory(body: { name: string; icon?: string }): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/categories`, body, { responseType: 'text' });
  }

  updateCategory(id: number, body: { name?: string; icon?: string }): Observable<string> {
    return this.http.put(`${this.apiUrl}/admin/categories/${id}`, body, { responseType: 'text' });
  }

  deleteCategory(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/admin/categories/${id}`, { responseType: 'text' });
  }

  /** Upload une icône (SVG, PNG, etc.) pour une catégorie. Retourne le chemin à mettre dans le champ icon. */
  uploadCategoryIcon(file: File): Observable<{ path: string }> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<{ path: string }>(`${this.apiUrl}/admin/categories/upload-icon`, formData);
  }

  // --- Gestion des Référentiels / Configuration ---

  // Options de livraison
  getAllDeliveryOptions(): Observable<DeliveryOptionDTO[]> {
    return this.http.get<DeliveryOptionDTO[]>(`${this.apiUrl}/admin/delivery-options`);
  }
  createDeliveryOption(dto: DeliveryOptionDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/delivery-options`, dto, { responseType: 'text' });
  }

  // Frais
  getAllFees(): Observable<FeeDTO[]> {
    return this.http.get<FeeDTO[]>(`${this.apiUrl}/admin/fees`);
  }
  createFee(dto: FeeDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/fees`, dto, { responseType: 'text' });
  }

  // Types de réclamation
  getAllClaimProblemTypes(): Observable<ReferenceItemDTO[]> {
    return this.http.get<ReferenceItemDTO[]>(`${this.apiUrl}/admin/claim-problem-types`);
  }
  createClaimProblemType(dto: ReferenceItemDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/claim-problem-types`, dto, { responseType: 'text' });
  }
  updateClaimProblemType(id: number, dto: ReferenceItemDTO): Observable<string> {
    return this.http.put(`${this.apiUrl}/admin/claim-problem-types/${id}`, dto, { responseType: 'text' });
  }
  deleteClaimProblemType(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/admin/claim-problem-types/${id}`, { responseType: 'text' });
  }

  // Raisons livreur (Motifs incidents livreur)
  getAllDeliveryIssueReasons(): Observable<ReferenceItemDTO[]> {
    return this.http.get<ReferenceItemDTO[]>(`${this.apiUrl}/admin/delivery-issue-reasons`);
  }
  createDeliveryIssueReason(dto: ReferenceItemDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/delivery-issue-reasons`, dto, { responseType: 'text' });
  }
  updateDeliveryIssueReason(id: number, dto: ReferenceItemDTO): Observable<string> {
    return this.http.put(`${this.apiUrl}/admin/delivery-issue-reasons/${id}`, dto, { responseType: 'text' });
  }
  deleteDeliveryIssueReason(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/admin/delivery-issue-reasons/${id}`, { responseType: 'text' });
  }

  // Raisons salarié (Motifs incidents salarié)
  getAllEmployeeDeliveryIssueReasons(): Observable<ReferenceItemDTO[]> {
    return this.http.get<ReferenceItemDTO[]>(`${this.apiUrl}/admin/employee-delivery-issue-reasons`);
  }
  createEmployeeDeliveryIssueReason(dto: ReferenceItemDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/employee-delivery-issue-reasons`, dto, { responseType: 'text' });
  }
  updateEmployeeDeliveryIssueReason(id: number, dto: ReferenceItemDTO): Observable<string> {
    return this.http.put(`${this.apiUrl}/admin/employee-delivery-issue-reasons/${id}`, dto, { responseType: 'text' });
  }
  deleteEmployeeDeliveryIssueReason(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/admin/employee-delivery-issue-reasons/${id}`, { responseType: 'text' });
  }

  // Secteurs d'activité
  getAllCompanySectors(): Observable<ReferenceItemDTO[]> {
    return this.http.get<ReferenceItemDTO[]>(`${this.apiUrl}/admin/company-sectors`);
  }
  createCompanySector(dto: ReferenceItemDTO): Observable<string> {
    return this.http.post(`${this.apiUrl}/admin/company-sectors`, dto, { responseType: 'text' });
  }
  updateCompanySector(id: number, dto: ReferenceItemDTO): Observable<string> {
    return this.http.put(`${this.apiUrl}/admin/company-sectors/${id}`, dto, { responseType: 'text' });
  }
  deleteCompanySector(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/admin/company-sectors/${id}`, { responseType: 'text' });
  }
}

export interface CategoryListItemDTO {
  id: number;
  name: string;
  icon?: string | null;
}
